### PR TITLE
Wip 308 add tests margin

### PR DESCRIPTION
**Pull request by LancelotMarti**
* Update to opendp 0.11.1
* Add margin tests
* Correct partitions logic
* Use datetime as string for covid synth (temp solution before opendp 0.12)

### CODE REVIEW

**Comment by PaulineMauryL** [comment]
## Coverage report


 <details><summary>Click to see where and how coverage changed</summary><table><thead>
  <tr><th>File</th><th>Statements</th><th>Missing</th><th>Coverage</th><th>Coverage<br>(new stmts)</th><th>Lines missing</th></tr>
</thead>
<tbody><tr>
<td colspan="6">&nbsp;&nbsp;<b>server/lomas_server</b></td><tr>
<td>&nbsp;&nbsp;<a href="https://github.com/dscc-admin-ch/lomas/pull/169/files#diff-4ef5cc30e18c28ff2745a2889f8b0ff899cca1f8bbf06fc33d0333dde1f48465">constants.py</a></td>
<td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/169/files#diff-4ef5cc30e18c28ff2745a2889f8b0ff899cca1f8bbf06fc33d0333dde1f48465"></a></td><td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/169/files#diff-4ef5cc30e18c28ff2745a2889f8b0ff899cca1f8bbf06fc33d0333dde1f48465"></a></td><td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/169/files#diff-4ef5cc30e18c28ff2745a2889f8b0ff899cca1f8bbf06fc33d0333dde1f48465"></a></td><td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/169/files#diff-4ef5cc30e18c28ff2745a2889f8b0ff899cca1f8bbf06fc33d0333dde1f48465"></a></td><td></td><tr>
<td colspan="6">&nbsp;&nbsp;<b>server/lomas_server/dp_queries/dp_libraries</b></td><tr>
<td>&nbsp;&nbsp;<a href="https://github.com/dscc-admin-ch/lomas/pull/169/files#diff-fa24a03009808c4b8ce9362d10c36fdcdaa1b1043361cfd8964094bee91c1ecb">opendp.py</a></td>
<td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/169/files#diff-fa24a03009808c4b8ce9362d10c36fdcdaa1b1043361cfd8964094bee91c1ecb"></a></td><td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/169/files#diff-fa24a03009808c4b8ce9362d10c36fdcdaa1b1043361cfd8964094bee91c1ecb"></a></td><td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/169/files#diff-fa24a03009808c4b8ce9362d10c36fdcdaa1b1043361cfd8964094bee91c1ecb"></a></td><td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/169/files#diff-fa24a03009808c4b8ce9362d10c36fdcdaa1b1043361cfd8964094bee91c1ecb"></a></td><td><a href="https://github.com/dscc-admin-ch/lomas/pull/169/files#diff-fa24a03009808c4b8ce9362d10c36fdcdaa1b1043361cfd8964094bee91c1ecbR46-R46">46</a></td><tr>
<td colspan="6">&nbsp;&nbsp;<b>server/lomas_server/private_dataset</b></td><tr>
<td>&nbsp;&nbsp;<a href="https://github.com/dscc-admin-ch/lomas/pull/169/files#diff-1e54c25951230aa65b4646aca131ee2ea874c4082887fcc6f1f924df6bd49d00">private_dataset.py</a></td>
<td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/169/files#diff-1e54c25951230aa65b4646aca131ee2ea874c4082887fcc6f1f924df6bd49d00"></a></td><td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/169/files#diff-1e54c25951230aa65b4646aca131ee2ea874c4082887fcc6f1f924df6bd49d00"></a></td><td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/169/files#diff-1e54c25951230aa65b4646aca131ee2ea874c4082887fcc6f1f924df6bd49d00"></a></td><td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/169/files#diff-1e54c25951230aa65b4646aca131ee2ea874c4082887fcc6f1f924df6bd49d00"></a></td><td></td><tr>
<td colspan="6">&nbsp;&nbsp;<b>server/lomas_server/tests</b></td><tr>
<td>&nbsp;&nbsp;<a href="https://github.com/dscc-admin-ch/lomas/pull/169/files#diff-d03a128b76f6654a4b0708c6b8f488eb7fc31178cbfcbec226e4d25bc88150de">test_api_opendp_polars.py</a></td>
<td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/169/files#diff-d03a128b76f6654a4b0708c6b8f488eb7fc31178cbfcbec226e4d25bc88150de"></a></td><td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/169/files#diff-d03a128b76f6654a4b0708c6b8f488eb7fc31178cbfcbec226e4d25bc88150de"></a></td><td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/169/files#diff-d03a128b76f6654a4b0708c6b8f488eb7fc31178cbfcbec226e4d25bc88150de"></a></td><td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/169/files#diff-d03a128b76f6654a4b0708c6b8f488eb7fc31178cbfcbec226e4d25bc88150de"></a></td><td></td><tr>
<td>&nbsp;&nbsp;<a href="https://github.com/dscc-admin-ch/lomas/pull/169/files#diff-37c83b1852cb948f86adcade97790a713b7067237ff614e9f26bf5c98087f605">test_margin_domain.py</a></td>
<td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/169/files#diff-37c83b1852cb948f86adcade97790a713b7067237ff614e9f26bf5c98087f605"></a></td><td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/169/files#diff-37c83b1852cb948f86adcade97790a713b7067237ff614e9f26bf5c98087f605"></a></td><td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/169/files#diff-37c83b1852cb948f86adcade97790a713b7067237ff614e9f26bf5c98087f605"></a></td><td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/169/files#diff-37c83b1852cb948f86adcade97790a713b7067237ff614e9f26bf5c98087f605"></a></td><td></td><tr>
<td colspan="6">&nbsp;&nbsp;<b>server/lomas_server/utils</b></td><tr>
<td>&nbsp;&nbsp;<a href="https://github.com/dscc-admin-ch/lomas/pull/169/files#diff-2fa5046d7b0bbe61fe4bb20a9e7eb645fd04ada9f9f725be9bb7fdc2642dd212">query_examples.py</a></td>
<td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/169/files#diff-2fa5046d7b0bbe61fe4bb20a9e7eb645fd04ada9f9f725be9bb7fdc2642dd212"></a></td><td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/169/files#diff-2fa5046d7b0bbe61fe4bb20a9e7eb645fd04ada9f9f725be9bb7fdc2642dd212"></a></td><td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/169/files#diff-2fa5046d7b0bbe61fe4bb20a9e7eb645fd04ada9f9f725be9bb7fdc2642dd212"></a></td><td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/169/files#diff-2fa5046d7b0bbe61fe4bb20a9e7eb645fd04ada9f9f725be9bb7fdc2642dd212"></a></td><td></td></tbody>
<tfoot>
<tr>
<td><b>Project Total</b></td>
<td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/169/files#diff-4b0bf2efa3367c0072ac2bf1e234e703dc46b47aaa4fe9d3b01737b1a15752b1"></a></td><td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/169/files#diff-4b0bf2efa3367c0072ac2bf1e234e703dc46b47aaa4fe9d3b01737b1a15752b1"></a></td><td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/169/files#diff-4b0bf2efa3367c0072ac2bf1e234e703dc46b47aaa4fe9d3b01737b1a15752b1"></a></td><td align="center"><a href="https://github.com/dscc-admin-ch/lomas/pull/169/files#diff-4b0bf2efa3367c0072ac2bf1e234e703dc46b47aaa4fe9d3b01737b1a15752b1"></a></td><td>&nbsp;</td>
</tr>
</tfoot>
</table>

<sub>

This report was generated by [python-coverage-comment-action](https://github.com/py-cov-action/python-coverage-comment-action)

</sub>
</details>

File: server/lomas_server/dp_queries/dp_libraries/opendp.py [hunk 1]
```diff
@@ -79,37 +81,34 @@ def get_lf_domain(metadata, by_config):
         )
         series_domains.append(series_domain)
 
-    lf_domain = dp.domains.lazyframe_domain(series_domains)
-
     # Margins
     # TODO Check lengths vs. keys for public info -> not in doc anymore.
 
     # Global params
-    params = {}
```
Comment (damienbfs) [inline server/lomas_server/dp_queries/dp_libraries/opendp.py:88]: Maybe rename params to margin_params?

File: server/lomas_server/dp_queries/dp_libraries/opendp.py [hunk 2]
```diff
@@ -79,37 +81,34 @@ def get_lf_domain(metadata, by_config):
...
-        lf_domain = dp.domains.with_margin(
-            lf_domain,
-            by=[],
-            public_info="lengths",
-            **params,
-        )
-
-    # Grouping logic (margin adaptation)
+    params = _get_global_params(metadata)
     if by_config:
```
Comment (damienbfs) [inline server/lomas_server/dp_queries/dp_libraries/opendp.py:89]: What if two group_by's in one query?

File: server/lomas_server/dp_queries/dp_libraries/opendp.py [hunk 3]
```diff
@@ -79,37 +81,34 @@ def get_lf_domain(metadata, by_config):
...
 
+def _get_global_params(metadata):
+    params = {}
+    params["max_num_partitions"] = 1
+    params["max_partition_length"] = metadata["rows"]
+
+    return params
+
+
 def _update_params_by_grouping(metadata, by_config, params):
```
Comment (damienbfs) [inline server/lomas_server/dp_queries/dp_libraries/opendp.py:112]: Not sure we need the underscore for these functions? get_lf_domain is also only used "internally" in this file for example. Only minor detail / matter of taste

File: server/lomas_server/tests/test_api_opendp_polars.py [hunk 4]
```diff
@@ -130,10 +130,47 @@ def test_opendp_polars_query(self) -> None:
             )
             assert response.status_code == status.HTTP_200_OK
 
+    def test_opendp_polars_datetime_query(self) -> None:
```
Comment (damienbfs) [inline server/lomas_server/tests/test_api_opendp_polars.py:133]: Add a todo to change this to datetime once we have opendp 1.12?

File: server/lomas_server/dp_queries/dp_libraries/opendp.py [hunk 5]
```diff
@@ -174,12 +172,13 @@ def _multiple_group_update_params(metadata, by_config, params):
 
     for column in by_config:
         series_info = metadata["columns"].get(column)
+        params["max_partition_length"] = min(
```
Comment (damienbfs) [inline server/lomas_server/dp_queries/dp_libraries/opendp.py:175]: Add comments about reasoning behind code?

**Review by damienbfs — APPROVED** [review]

File: server/lomas_server/dp_queries/dp_libraries/opendp.py (hunk 3 above)
Comment (LancelotMarti) [reply server/lomas_server/dp_queries/dp_libraries/opendp.py:112]: make sense

File: server/lomas_server/dp_queries/dp_libraries/opendp.py (hunk 1 above)
Comment (LancelotMarti) [reply server/lomas_server/dp_queries/dp_libraries/opendp.py:88]: done

File: server/lomas_server/dp_queries/dp_libraries/opendp.py (hunk 5 above)
Comment (LancelotMarti) [reply server/lomas_server/dp_queries/dp_libraries/opendp.py]: done

**Pull request merged**